### PR TITLE
gdm 1.0 (new formula)

### PR DESCRIPTION
**Pull request by sparrc**

### CODE REVIEW

**Comment by sparrc** [comment]
I have a problem: `gdm` depends on the `go` executable to work, is there a way I can access the `go` command within the `test do` block?

**Comment by mikemcquaid** [comment]
@sparrc If it depends on `go` it's not a build-time dependency.

File: Library/Formula/gdm.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+             "-ldflags", "-X main.Version=#{version}"
+    end
+
+    bin.install gdm_path/"gdm"
+  end
+
+  test do
+    cd testpath do
+      ENV["GOPATH"] = testpath
+      assert_match "#{version}", shell_output("#{bin}/gdm version")
```
Comment (mikemcquaid) [inline Library/Formula/gdm.rb]: Would be good if you could actually generate some `go_resource`s in the test.

File: Library/Formula/gdm.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  go_resource "golang.org/x/tools" do
+    url "https://go.googlesource.com/tools.git",
+    :revision => "b48dc8da98ae78c3d11f220e7d327304c84e623a"
+  end
+
+  def install
+    ENV["GOPATH"] = buildpath
+    gdm_path = buildpath/"src/github.com/sparrc/gdm"
+    gdm_path.install Dir["*"]
```
Comment (mikemcquaid) [inline Library/Formula/gdm.rb]: Can you make a symlink instead? Will be a bit faster.

File: Library/Formula/gdm.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    Language::Go.stage_deps resources, buildpath/"src"
+
+    cd gdm_path do
+      system "go", "build", "-o", "gdm",
+             "-ldflags", "-X main.Version=#{version}"
+    end
+
+    bin.install gdm_path/"gdm"
```
Comment (bfontaine) [inline Library/Formula/gdm.rb]: You might want to move this into the `cd gdm_path` block and use `bin.install "gdm"`.

File: Library/Formula/gdm.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  test do
+    cd testpath do
```
Comment (bfontaine) [inline Library/Formula/gdm.rb]: You don’t need to `cd` into the `testpath`; you’re already into it :wink:

File: Library/Formula/gdm.rb (hunk 1 above)
Comment (sparrc) [inline Library/Formula/gdm.rb]: That's going to be difficult, gdm can only generate those if it's within a legit Go project. Is there a way that I can access `build_path` within the `test do` block?

File: Library/Formula/gdm.rb (hunk 2 above)
Comment (sparrc) [inline Library/Formula/gdm.rb]: Tried to do a symlink but I got this when building:

```console
==> go build -o gdm -ldflags -X main.Version=1.0
Last 15 lines from /Users/csparr/Library/Logs/Homebrew/gdm/01.go:
2015-12-11 10:16:16 -0700

go
build
-o
gdm
-ldflags
-X main.Version=1.0

can't load package: package github.com/sparrc/gdm: open /private/tmp/gdm20151211-1601-1lb2nr/gdm-1.0/src/github.com/sparrc/gdm/imports.go: too many levels of symbolic links
```

File: Library/Formula/gdm.rb (hunk 3 above)
Comment (sparrc) [inline Library/Formula/gdm.rb]: done

File: Library/Formula/gdm.rb (hunk 4 above)
Comment (sparrc) [inline Library/Formula/gdm.rb]: done

File: Library/Formula/gdm.rb (hunk 2 above)
Comment (mikemcquaid) [inline Library/Formula/gdm.rb]: What was the symlink you did? Maybe needs to be the reverse.

File: Library/Formula/gdm.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+             "-ldflags", "-X main.Version=#{version}"
+      bin.install "gdm"
+    end
+  end
+
+  test do
+    ENV["GOPATH"] = testpath
+    assert_match "#{version}", shell_output("#{bin}/gdm version")
+    assert_match "#{testpath}", shell_output("#{bin}/gdm save")
+    system "#{bin}/gdm", "restore"
```
Comment (mikemcquaid) [inline Library/Formula/gdm.rb]: :+1:

File: Library/Formula/gdm.rb (hunk 2 above)
Comment (sparrc) [inline Library/Formula/gdm.rb]: I just changed 

```
gdm_path.install Dir["*"]
```

to
 
```
gdm_path.install_symlink Dir["*"]
```

**Comment by sparrc** [comment]
Looks like my repository isn't notable enough? :sweat_smile:

```
==> audit problems
gdm:
 * GitHub repository not notable enough (<10 forks, <10 watchers and <20 stars)
 * GitHub repository too new (<30 days old)
```

**Comment by mikemcquaid** [comment]
@sparrc Usually yep but if we decide to link this from the documentation we may make an exception. We may need to have more discussion there, though.

**Comment by sparrc** [comment]
@mikemcquaid Fair enough, I @'ed the maintainer of homebrew-go-resources in the doc discussion on the doc change: #46889 so we can get that settled :grin:

**Comment by DomT4** [comment]
Anything left to be done here?

**Comment by sparrc** [comment]
@DomT4 Nope, if it's okay to waive the audit then it's ready to merge

**Comment by sparrc** [comment]
Actually looks like I might want to symlink instead of `gdm_path.install Dir["*"]`, @mikemcquaid I wasn't quite able to get that working, do you have any suggestions?

**Comment by DomT4** [comment]
If the symlink method doesn't work I'm not wedded to us trying to force it, personally, but I'll defer to @mikemcquaid since he raised the question originally.

**Comment by mikemcquaid** [comment]
@sparrc Check out the e.g. `godep` formula. If that approach doesn't work it might be worth figuring out what makes your project different from a standard Go project.

**Comment by DomT4** [comment]
test failed this time:
```
==> /usr/local/Cellar/gdm/1.0/bin/gdm version
==> gdm save
Error running go list -f {{range .Deps}}{{.}}{{"\n"}}{{end}} ./... in dir /private/tmp/gdm20151223-71769-17f3csw, exec: "go": executable file not found in $PATH
Error: gdm: failed
<0> expected but was
<1>.
```

File: Library/Formula/gdm.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    Language::Go.stage_deps resources, buildpath/"src"
+
+    cd "src/github.com/sparrc/gdm" do
+      system "go", "build", "-o", bin/"gdm",
+             "-ldflags", "-X main.Version=#{version}"
+    end
+  end
+
+  test do
+    cd testpath do
```
Comment (DomT4) [inline Library/Formula/gdm.rb]: You're already in the `testpath` here by the way. Don't need to `cd` into it.

**Comment by sparrc** [comment]
@mikemcquaid @DomT4 I believe this is ready now, I made it symlink and as far as I can tell it's only  failing the audit check (for repo not being notable enough)

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Pull request closed without merging**